feat: when using replace strategy also update PR

dnwe opened the pull request:
# What does this change

Previous when using `--conflict-strategy replace` multi-gitter would replace all of the commits, but would leave the pull request itself untouched if it was already open.

Now if multi-gitter is re-run and (e.g.,) the pr-title, pr-body or labels are changed, then the existing PR will be edited and updated to use their new values

# Checklist
- [x] Made sure the PR follows the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines
- [x] Tests if something new is added

lindell commented on internal/multigitter/run.go:
```diff
@@ -323,8 +326,26 @@ func (r *Runner) runSingleRepo(ctx context.Context, repo scm.Repository) (scm.Pu
+				Body:          r.PullRequestBody,
+				Head:          r.FeatureBranch,
+				Base:          baseBranch,
+				Reviewers:     getReviewers(r.Reviewers, r.MaxReviewers),
+				TeamReviewers: getReviewers(r.TeamReviewers, r.MaxTeamReviewers),
```
Thought/Discussion: If `max-reviewers` is less than the reviewers provided, the reviewers used will be randomized. Do we want this to happen in the case of an update or not? 🤔 I have no real opinion here, but a conscious decision has to be made.

lindell reviewed (commented):
Thanks for the PR!

The code looks really good! I don't really have any code comments other than that we might be able to share some code between the Create and Update functions. But it is probably also as much of premature optimization, so I did not actually put in any comments on it. 

----

The thing that needs to be discussed is what should happen to fields such as reviewers/labels/etc.

It seems that the behavior differs between SCMs (and for Giteas case, per field type). In some cases, the fields are updated to the list of the values provided to multi-gitter, while in other cases, the new values are added to any existing ones. 

* Gitlab seems to replace the labels https://github.com/lindell/multi-gitter/blob/922787afec238e288196597691d11c1126d877c1/internal/scm/gitlab/gitlab.go#L329
* While GitHub adds the new ones https://github.com/lindell/multi-gitter/blob/922787afec238e288196597691d11c1126d877c1/internal/scm/github/github.go#L398

I think the replacement approach seems more logical. But I'm happy to discuss which one should be used. Either way, it should be the same no matter of SCM/field type.

jamestelfer commented:
This is a pretty useful PR that would help me out -- what is needed to get it over the finish line? Could I resurrect it to get that last bit done?

dnwe commented:
@jamestelfer thanks for the prod, I'd forgotten to revisit this PR after the review from @lindell — let me do a quick rebase and then take a look at addressing the comments

lindell commented:
I now commit a change for the "update instead of add" change for github. I think that only Gitea reviewers are left now.

dnwe commented:
@lindell I've rebased on later master and I think this PR should be ready for a re-review now

dnwe commented:
Note: for gitea, I spun up a docker image with an install and configured three sample users and tested that the replace of review requests was working 

<img width="300" alt="image" src="https://github.com/lindell/multi-gitter/assets/8060970/76cf98af-b152-4199-be20-2f73bffd401e">

lindell commented on internal/scm/gitea/gitea.go:
```diff
@@ -305,6 +307,85 @@ func (g *Gitea) getLabelsFromStrings(ctx context.Context, repo repository, label
+	r := repo.(repository)
+
+	var addedReviewers, removedReviewers []string
+	if newPR.Reviewers != nil {
```
Return early if `newPR.Reviewers == nil` instead

lindell commented on internal/scm/gitea/gitea.go:
```diff
@@ -305,6 +307,85 @@ func (g *Gitea) getLabelsFromStrings(ctx context.Context, repo repository, label
 }
 
+func (g *Gitea) setReviewers(ctx context.Context, repo scm.Repository, newPR scm.NewPullRequest, createdPR *gitea.PullRequest) error {
+	r := repo.(repository)
```
Use `repository` directly in the argument instead since this is not a public method.

lindell commented on internal/scm/gitea/gitea.go:
```diff
@@ -7,6 +7,8 @@ import (
 	"sort"
 	"strings"
 
+	"golang.org/x/exp/slices"
```
slices.DeleteFunc now exist in the std slices package now 😄

dnwe replied:
Yeah happy to switch that over. When I coded this up it was before I'd rebased the old branch so I hadn't picked up #402 and thought you were still on go1.19 as a baseline.

FYI scm/github/github.go is also using golang.org/x/exp/maps which is available as just maps now too

dnwe replied:
(I'll update both under this PR)

dnwe replied:
Ah my mistake, looks like [maps.Values](https://pkg.go.dev/golang.org/x/exp/maps#Values) in exp/maps is newer than the stdlib maps — even 1.22 doesn't have it so it probably won't come until go1.23

lindell reviewed (commented):
Just some minor comments and this should be ready to me merged 😄

lindell reviewed (approved):
LGTM 👍 

Thanks for the contribution!
